client/rpc/src/state/state_full.rs

```diff
@@ -434,176 +420,175 @@ where
+			.runtime_version_at(&BlockId::hash(client.info().best_hash))
+			.expect("best hash is valid; qed");
+		let _ = sink.send(&previous_version);
+		let rt_version_stream = client
+			.storage_changes_notification_stream(
+				Some(&[StorageKey(well_known_keys::CODE.to_vec())]),
+				None,
+			)
+			.map_err(|blockchain_err| Error::Client(Box::new(blockchain_err)))?;
+
+		let fut = async move {
+			rt_version_stream
+				.filter_map(|_| {
+					let info = client.info();
+					let version = client.runtime_version_at(&BlockId::hash(info.best_hash));
+					match version {
+						Ok(v) =>
+							if previous_version != v {
+								previous_version = v.clone();
+								future::ready(Some(v))
+							} else {
+								future::ready(None)
+							},
+						Err(e) => {
+							log::error!("Could not fetch current runtime version. Error={:?}", e);
+							future::ready(None)
+						},
+					}
+				})
+				.take_while(|version| {
+					future::ready(sink.send(&version).map_or_else(
+						|e| {
+							log::error!("Could not send data to the state_subscribeRuntimeVersion subscriber: {:?}", e);
+							false
+						},
+						|_| true,
+					))
+				})
+				.for_each(|_| future::ready(()))
+				.await;
+		}
+		.boxed();
+		executor.execute(fut);
+
+		Ok(())
 	}
 
 	fn subscribe_storage(
 		&self,
-		_meta: crate::Metadata,
-		subscriber: Subscriber<StorageChangeSet<Block::Hash>>,
+		mut sink: SubscriptionSink,
 		keys: Option<Vec<StorageKey>>,
-	) {
-		let keys = Into::<Option<Vec<_>>>::into(keys);
-		let stream = match self
-			.client
-			.storage_changes_notification_stream(keys.as_ref().map(|x| &**x), None)
-		{
-			Ok(stream) => stream,
-			Err(err) => {
-				let _ = subscriber.reject(client_err(err).into());
-				return
-			},
-		};
-
-		// initial values
-		let initial = stream::iter(
-			keys.map(|keys| {
-				let block = self.client.info().best_hash;
-				let changes = keys
-					.into_iter()
-					.map(|key| {
-						let v = self.client.storage(&BlockId::Hash(block), &key).ok().flatten();
-						(key, v)
+	) -> std::result::Result<(), Error> {
+		let executor = self.executor.clone();
+		let client = self.client.clone();
+
+		let stream = client
+			.storage_changes_notification_stream(keys.as_ref().map(|keys| &**keys), None)
+			.map_err(|blockchain_err| Error::Client(Box::new(blockchain_err)))?;
+
+		let block = client.info().best_hash;
+		let changes: Vec<(StorageKey, Option<StorageData>)> = keys
+			.map(|keys| {
+				keys.into_iter()
+					.map(|storage_key| {
+						let v = client.storage(&BlockId::Hash(block), &storage_key).ok().flatten();
+						(storage_key, v)
 					})
-					.collect();
-				vec![Ok(Ok(StorageChangeSet { block, changes }))]
+					.collect()
 			})
-			.unwrap_or_default(),
-		);
-
-		self.subscriptions.add(subscriber, |sink| {
-			let stream = stream.map(|(block, changes)| {
-				Ok(Ok::<_, rpc::Error>(StorageChangeSet {
-					block,
-					changes: changes
+			.unwrap_or_default();
+		if !changes.is_empty() {
```

Comment:
@dvdplm I added this check to not send `empty StorageChangeSet` because there were some issues in the `polkadot UI` and it doesn't make sense to send them anyway because they are useless anyway.

For context this may happen if the `storage keys` is empty but I think it's a timing issue where `master` is doing this in executor and we are actually blocking the current thread for the initial `StorageChangeSet`